### PR TITLE
Make ballot unique

### LINKED ISSUE #1346: Ballots can't be unique on (post,election)
I added the four South Lakeland by-elections to EE based on https://www.southlakeland.gov.uk/voting-and-elections/elections/casual-vacancies-and-postponed-elections/#gsc.tab=0

So there are now new IDs in EE for the by-elections, which differ from the cancelled scheduled elections:
`local.south-lakeland.broughton-coniston.2021-05-06` (cancelled)
`local.south-lakeland.broughton-coniston.by.2021-05-06` (by-election)

The election group `local.south-lakeland.2021-05-06` no longer shows as cancelled on https://elections.democracyclub.org.uk/elections/local.2021-05-06/ whereas the other 5 cancelled elections do (Cumbria, North Yorkshire etc.)

But the by-election IDs are not showing on https://candidates.democracyclub.org.uk/elections/local.south-lakeland.2021-05-06/ and can't be selected on candidate pages
image

### PR DESCRIPTION
**Pull request by symroe**
Across the project we have made the assumption that (election, post) would make a ballot unique.

This is a migration from the initial version of the site that didn't use the concept of a ManyToMany between `Election` and `Post` at all. We added the join table in 2016, converted it to be called `Ballot` in ~2018, but there are still a lot of assumptions based on the `(election, post)` combo.

This was fine to ignore as some tech debt until #1346 came along, and forced the issue.

This PR does the work of converting `Ballot.ballot_paper_id` to be the only unique constraint. We have this constraint in EE, so it shouldn't cause any problems.

Still to do:

* [x] Add Unique constraint on `Ballot.ballot_paper_id`
* [x] Add tests to the EE importer to check it can create ballots with duplicate `(election, post)` FKs
* [x] Try to find other instances where we assume `election, post` in e.g lookups. This might be hard to do 100%, so we'll have to keep an eye on errors relating to this once merged and deployed.

Fixes #1346

### CODE REVIEW

File: ynr/apps/elections/templates/elections/party_for_ballot.html [hunk 1]
```diff
@@ -37,7 +37,7 @@ <h1>
                 {% if DATE_TODAY >= election.election_date %}
                   {% if user_can_record_results and show_confirm_result and not candidate_elected %}
                     {% if 'local' not in election.slug %}
-                      <form class="winner-confirm" action="{% url 'record-winner' election=election.slug post_id=post_data.id %}" method="post">
+                      <form class="winner-confirm" action="{% url '(?P<election>[^/]+)' election=election.slug post_id=post_data.id %}" method="post">
```
Comment (michaeljcollinsuk) [inline ynr/apps/elections/templates/elections/party_for_ballot.html:40]: Just checking this is correct, as it is different to the other url changes?

Comment (symroe) [reply ynr/apps/elections/templates/elections/party_for_ballot.html]: Good spot, this must have been a find and replace error!

**Review by michaeljcollinsuk — APPROVED** [review]
LGTM. Couple of minor things but not really important for now
- some of the urls have trailing slashes, and some don't, might be worth making this consistent at some point?
- again related to urls - at some point would be nice to refactor them to use `path`. Partly just because I always find regexs confusing! But again not important for now, and wouldn't want to risk breaking anything
